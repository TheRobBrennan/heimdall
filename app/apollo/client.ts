import fetch from "cross-fetch"
import { useMemo } from "react"

import { ApolloClient, createHttpLink, InMemoryCache } from "@apollo/client"
import { setContext } from "@apollo/client/link/context"

let apolloClient

const httpLink = createHttpLink({
  uri: `api/graphql`,
  fetch,
})

/* istanbul ignore next */
const authLink = setContext(async () => {
  // See https://www.apollographql.com/docs/react/networking/authentication/#header for more details
  // Get our session detail from the back-end API
  let authorization = ""
  try {
    const data = await (await fetch("api/session")).json()
    const { accessToken } = data
    authorization = `Bearer ${accessToken}`
  } catch (e) {
    // No session data available; no authorization header will be sent
  }

  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      authorization,
    },
  }
})

function createApolloClient(initialState) {
  const ssrMode = typeof window === "undefined"

  return new ApolloClient({
    ssrMode,
    link: authLink.concat(httpLink),
    cache: new InMemoryCache().restore(initialState),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient(initialState)

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  /* istanbul ignore next */
  if (typeof window === "undefined") return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

/* istanbul ignore next */
export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
