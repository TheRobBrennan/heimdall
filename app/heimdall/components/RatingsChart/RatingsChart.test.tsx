import * as React from "react"
import { cleanup, render } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"

import Component, { GET_DATA_QUERY } from "./RatingsChart"

describe("GRANDstack RatingsChart component", () => {
  afterEach(cleanup)

  describe("when invoked WITHOUT a specific height or size", () => {
    it("should render the Ratings Distribution chart after receiving data", () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        result: {
          data: {
            ratingsCount: [
              { stars: 3, count: 2, __typename: "RatingCount" },
              { stars: 4, count: 4, __typename: "RatingCount" },
              { stars: 5, count: 6, __typename: "RatingCount" },
            ],
          },
        },
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      // Test
      const { asFragment } = render(
        <MockedProvider mocks={gqlMocks} addTypename={false}>
          <Component />
        </MockedProvider>
      )

      expect(asFragment).toMatchSnapshot()

      // TODO: Advance to the next tick in the event loop so our chart can render
      // await act(() => {
      //   return new Promise((resolve) => {
      //     setTimeout(resolve, 0)
      //   })
      // })

      // Verify
    })
  })

  describe("when invoked WITH a specific height and size", () => {
    it("should render the Ratings Distribution chart after receiving data", () => {
      // Define our Apollo request
      const renderRequest = {
        request: {
          query: GET_DATA_QUERY,
          variables: {},
        },
        result: {
          data: {
            ratingsCount: [
              { stars: 3, count: 2, __typename: "RatingCount" },
              { stars: 4, count: 4, __typename: "RatingCount" },
              { stars: 5, count: 6, __typename: "RatingCount" },
            ],
          },
        },
      }

      // Define our mock response(s)
      const gqlMocks = [renderRequest]

      // Test
      const { asFragment } = render(
        <MockedProvider mocks={gqlMocks} addTypename={false}>
          <Component />
        </MockedProvider>
      )

      expect(asFragment).toMatchSnapshot()

      // TODO: Advance to the next tick in the event loop so our chart can render
      // await act(() => {
      //   return new Promise((resolve) => {
      //     setTimeout(resolve, 0)
      //   })
      // })

      // Verify
    })
  })
})
