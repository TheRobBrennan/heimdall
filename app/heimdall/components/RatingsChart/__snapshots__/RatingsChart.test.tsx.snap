// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GRANDstack RatingsChart component when invoked WITH a specific height and size should render the Ratings Distribution chart after receiving data 1`] = `[Function]`;

exports[`GRANDstack RatingsChart component when invoked WITHOUT a specific height or size should render the Ratings Distribution chart after receiving data 1`] = `[Function]`;
